contracts/gateway/StakerProxy.sol

```diff
@@ -0,0 +1,200 @@
+ * @notice The composer contract will deploy StakerProxy contracts on behalf of
+ *         the owner. Only the owner can self-destruct the StakerProxy and
+ *         receive all stored value.
+ */
+contract StakerProxy {
+
+    /* Storage */
+
+    /** The composer that deployed this contract. */
+    ComposerInterface public composer;
+
+    /** The composer deployed the StakerProxy on behalf of the owner. */
+    address payable public owner;
+
+    /** A mutex to prevent reentrancy from a malicious gateway. */
+    bool stakeMutex;
+
+
+    /* Modifiers */
+
+    /** Requires that msg.sender is the composer */
+    modifier onlyComposer() {
+        require(
+            msg.sender == address(composer),
+            "This function can only be called by the composer."
+        );
+
+        _;
+    }
+
+    /** Requires that msg.sender is the owner. */
+    modifier onlyOwner() {
+        require(
+            msg.sender == owner,
+            "This function can only be called by the owner."
+        );
+
+        _;
+    }
+
+    /** Mutex to prevent reentrancy attack. */
+    modifier mutexed() {
+        assert(!stakeMutex);
+        stakeMutex = true;
+
+        _;
+
+        stakeMutex = false;
+    }
+
+
+    /* Special Functions */
+
+    /**
+     * @notice Must be constructed by a contract that implements the
+     *         `ComposerInterface`.
+     *
+     * @param _owner The owner that this proxy is deployed for.
+     */
+    constructor(address payable _owner) public {
+        composer = ComposerInterface(msg.sender);
+        owner = _owner;
+    }
+
+
+    /* External Functions */
+
+    /**
+     * @notice Initiates the stake process. In order to stake, the stake and
+     *         bounty amounts must first be transferred to the StakerProxy.
+     *         Staked amount is then transferred to the Gateway contract.
+     *         Bounty amount is also transferred to the Gateway contract.
+     *
+     * @param _amount Stake amount that will be transferred.
+     * @param _beneficiary The address in the auxiliary chain where the utility
+     *                     tokens will be minted.
+     * @param _gasPrice Gas price that staker is ready to pay to get the stake
+     *                  and mint process done.
+     * @param _gasLimit Gas limit that staker is ready to pay.
+     * @param _nonce The nonce to verify it is as expected.
+     * @param _hashLock Hash lock provided by the facilitator.
+     * @param _gateway On which gateway to stake.
+     *
+     * @return messageHash_ Message hash is unique for each request.
+     */
+    function stake(
+        uint256 _amount,
+        address _beneficiary,
+        uint256 _gasPrice,
+        uint256 _gasLimit,
+        uint256 _nonce,
+        bytes32 _hashLock,
+        EIP20GatewayInterface _gateway
+    )
+        external
+        onlyComposer
+        mutexed
+        returns (bytes32 messageHash_)
+    {
+        approveTransfers(_gateway, _amount);
```

Comment:
We can do `approveTransfers` after the nonce validations

Reply:
Will move.